app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+            }
+            photoSaveSuccess &= FileIO.saveBitmap(bitmaps.get(i), bitmapPath);
+        }
+        return photoSaveSuccess;
+    }
+
+    /**
+     * This method writes the ad panoramas on disk. The actual operation is
+     * performed by the writeImage function.
+     *
+     * @param panoramas list of bitmaps that represents the ad panoramas
+     * @return a boolean that indicates if the operation succeeded or not
+     */
+    private boolean writePanoramas(List<Bitmap> panoramas) {
+        return writeImage(panoramas,
+                ImageType.PANORAMA);
+    }
+
+    /**
+     * This method writes the profile picture of a user on disk. Actually,
+     * the write on disk is performed by loadProfilePic. The correct path is
+     * given to this function. We opted to do this, this way because we do
+     * not want the local database to be aware if Firebase.
+     *
+     * @param loadProfilePic the function that performs the write on disk of
+     *                       the profile picture of the user
+     * @param futureSuccess  a completable future that indicates if the
+     *                       operation succeeded or not.
+     */
+    private void writeProfilePic(Function<String,
+            CompletableFuture<Void>> loadProfilePic,
+                                 CompletableFuture<Void> futureSuccess) {
+
+        CompletableFuture<Void> futureProfilePic =
+                loadProfilePic.apply(LocalDatabasePaths.userProfilePic(getCurrentUser().getUserId()));
+        futureProfilePic.thenAccept(arg -> futureSuccess.complete(null));
+        futureProfilePic.exceptionally(e -> {
+            futureSuccess.completeExceptionally(e);
+            return null;
+        });
+
+        futureSuccess.complete(null);
+    }
+
+    /**
+     * This function performs the writing of a complete ad into local storage
+     * . It will create a folder for the added user only if it doesn't
+     * already exists (in which case it is updated). It will do the same for
+     * the ad. It also performs the writing of all the images. The photos of
+     * ad, the panoramas of the ad and the profile picture of the user (if he
+     * doesn't uses the default one).
+     *
+     * @param adId           the id of the ad
+     * @param cardId         the id of the card
+     * @param ad             the ad
+     * @param user           the user who posted the ad
+     * @param adPhotos       list of bitmaps representing the photos of the ad
+     * @param panoramas      list of panoramas representing the panoramas of
+     *                       the ad
+     * @param loadProfilePic the function that will perform the writing on
+     *                       disk of the user's profile picture. It receives
+     *                       the path representing the location of the
+     *                       profile picture and returns a completable future
+     *                       to indicate if the operation succeeded or not.
```

Comment:
Oh nice, I didn't know `GlideBitmapLoader ` could directly get the images from Firestore. This will make the process of loading images more unified and simple.

Reply:
Yes it can baby !

Reply:
But there is a but, I think it requires context... Which I don't think is available here ? Or, in fact, its okay if this class gets injected with hilt into an activity.

Reply:
to be honest I realized it while I was reviewing, otherwise I would have suggested it already during our previous discussions, sorry